internal/cmd/local/airbyte/api.go

```diff
@@ -0,0 +1,235 @@
+		return fmt.Errorf("failed to update organization: %w", err)
+	}
+
+	return nil
+}
+
+// getOrg returns the default organization "00000000-0000-0000-0000-000000000000".
+func (a *Airbyte) getOrg(ctx context.Context) (organization, error) {
+	token, err := a.fetchToken(ctx)
+	if err != nil {
+		return organization{}, fmt.Errorf("unable to fetch token: %w", err)
+	}
+
+	type orgReq struct {
+		OrgID string `json:"organizationId"`
+	}
+
+	jsonData, err := json.Marshal(orgReq{OrgID: "00000000-0000-0000-0000-000000000000"})
+	if err != nil {
+		return organization{}, fmt.Errorf("unable to marshal organization request: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, a.host+pathOrgGet, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return organization{}, fmt.Errorf("unable to create organization request: %w", err)
+	}
+	req.Header.Add("content-type", "application/json")
+	req.Header.Add("accept", "application/json")
+	req.Header.Add("Authorization", "Bearer "+string(token))
+
+	res, err := a.h.Do(req)
+	if err != nil {
+		fmt.Errorf("unable to send organization request: %w", err)
+	}
+
+	resBody, err := io.ReadAll(res.Body)
+	if err != nil {
+		return organization{}, fmt.Errorf("unable to read organization response: %w", err)
+	}
+
+	var org organization
+	if err := json.Unmarshal(resBody, &org); err != nil {
+		return organization{}, fmt.Errorf("unable to decode organization request: %w", err)
+	}
+	return org, nil
+}
+
+// setOrg calls the organization/update endpoint.
+// This is a POST endpoint and does not support PATCH operations.
+// Make sure the org provides has all of its attributes defined.
+func (a *Airbyte) setOrg(ctx context.Context, org organization) error {
+	token, err := a.fetchToken(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to fetch token: %w", err)
+	}
+
+	jsonData, err := json.Marshal(org)
+	if err != nil {
+		return fmt.Errorf("unable to marshal organization request: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, a.host+pathOrgSet, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return fmt.Errorf("unable to create organization request: %w", err)
+	}
+	req.Header.Add("content-type", "application/json")
+	req.Header.Add("accept", "application/json")
+	req.Header.Add("Authorization", "Bearer "+string(token))
+
+	res, err := a.h.Do(req)
+	if err != nil {
+		fmt.Errorf("unable to send organization request: %w", err)
+	}
+	_ = res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status code: %d", res.StatusCode)
+	}
+
+	return nil
+}
+
+// fetchToken returns the application token for the ClientID and ClientSecret.
+// The token will be cached when first called, returning the cached value for every following call.
+func (a *Airbyte) fetchToken(ctx context.Context) (Token, error) {
+	t := a.token
+	if t != "" {
+		return t, nil
+	}
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	// check again if there is now a token
```

Comment:
I'm not following the reason for this check-lock-check-again behavior? Is this intended or leftover from some refactor?

Reply:
It's intended.  If two callers to `fetchToken` come in at the same time, the first would get the lock, create the token, and release the lock, the second would then get the lock, create a new token, and release the lock.  This prevents the second caller from recreating the token when the first caller already succeeded.

In the current configuration this could be replaced by a `sync.Once` to prevent this, but tokens don't last forever and I'm unsure when a new one will need to be fetched.  So the `token` itself isn't immutable.